### PR TITLE
auto calculate showamount, fix #106

### LINKED ISSUE #106: Playlist is overflowing mpv window
I don't see what I'm choosing after 11 lines. I hope you got what I meant

### PR DESCRIPTION
**Pull request by verygoodlee**
### Some test cases, only list key options
- `scale_playlist_by_window=no` `showamount=auto` `style_ass_tags={}` `text_padding_y=10`

- `scale_playlist_by_window=no` `showamount=auto` `style_ass_tags={\fs15}` `text_padding_y=10`
{fs15} 10
- `scale_playlist_by_window=no`  `showamount=auto` `style_ass_tags={\fs15}` `text_padding_y=30`
{fs15} 30
- `scale_playlist_by_window=no` `showamount=auto` `style_ass_tags={\fs30}` `text_padding_y=10`
{fs30} 10

### Known bugs
1. If options like this, the font size of each line is different from the global `style_ass_tags`, auto calculate is based on `style_ass_tags`'s font size, so it must have a bug, but I think nobody would set this strange configuration, this bug can be ignored
```
style_ass_tags={\fs15}
playlist_header={\fs20}Playlist [%cursor/%plen]
normal_file={\fs16}○ %name
hovered_file={\fs16}● %name
playing_hovered_file={\fs18}▶ %name
...
```
2. **[fixed]** ~~If some long file names occupy more than one line, it will definitely overflow, I have no idea to get the number of filename occupied lines. It's better to use  `showamount=auto`  and `slice_longfilenames=yes` together.~~
3. **[fixed]** ~~`scale_playlist_by_window=no` is required, at present, I don't know how `scale_playlist_by_window=yes` works~~

### CODE REVIEW

File: playlistmanager.lua [hunk 1]
```diff
@@ -237,6 +238,40 @@ if settings.system=="auto" then
...
+      fs = tonumber(ass_fs_tag:match('%d+'))
+    end
+  end
+ 
+  settings.showamount = math.floor(playlist_h / fs)
+  
+  -- exclude the header line
+  if settings.playlist_header ~= "" then
+    settings.showamount = settings.showamount - 1
+  end
```
Comment (jonniek) [inline playlistmanager.lua]: I think here we should probably take newlines into account here.

Alternatively we could have a setting where you can define some N amount that is always subtracted from the auto showamount. That could be a bit easier since you could also adjust for different font sizes etc. But it does a little bit defeat the purpose of auto setting.

**Review by jonniek — COMMENTED** [review]
Thanks for the PR. I'll have to do some manual testing for this. On a quick look it seems quite good, maybe some settings defaults to be adjusted but I can also do that later.

I think from my limited testing the file being on two lines is kind of an issue. I think the wrapping needs to be solved first. I think it's outside the scope of this PR though. I'll take a look at it.

File: playlistmanager.lua (hunk 1 above)
Comment (verygoodlee) [reply playlistmanager.lua]: `showamount`  subtract the number of `%N` and `\N` is very easy, but the height of empty line is much smaller than non empty line，which can leave too much blank space at the bottom.\
For Example `playlist_header=Playlist [%cursor/%plen]%N%N%N%N` ends with 4 empty lines, but the height is about 2 non empty lines.
image
I modified the `parse_header()` function to add a blank character at the end of each `\N`, problem seems to be solved.
image

**Comment by verygoodlee** [comment]
In `scale_playlist_by_window=yes` case,  there is still no progress. 
https://github.com/jonniek/mpv-playlistmanager/blob/7b8292589212a5bac1f312a555ec20aa561f972a/playlistmanager.lua#L671-L672
`mp.set_osd_ass(0, 0, ass.text)`, how does it work, I can't find it in the  [Reference Manual](https://mpv.io/manual/stable/), @jonniek do you have any documents about it?   If not, may need to find the answer in mpv source code

**Comment by jonniek** [comment]
Here is the source https://github.com/mpv-player/mpv/blob/e22a2f04833852ce825eb7c1235d9bdaaa9b2397/player/lua/defaults.lua#L688

I think it was deprecated by https://mpv.io/manual/stable/#lua-scripting-mp-create-osd-overlay(format)

The overlay looks like this

```lua
 {"res_x" = 0, "format" = "ass-events", "res_y" = 720, "id" = 1, "data" = ""}
```

So I guess it's possible to mutate the resolution and update:

```lua
local ov = mp.create_osd_overlay("ass-events")
ov.data = "Hello world"
ov.res_x = 0
ov.res_y = 0
ov:update()
```
But it doesn't seem to have the effect of scaling the text. Seems like this setting has probably been broken for a long time. I think we can remove this setting and move to using the overlay API that is documented (not touch the size). I think I tried to implement it before at some point but there was some issue with the sizing but I can't remember exactly.

So probably it should be something like:

```lua
local playlist_osd_overlay = mp.create_osd_overlay("ass-events")

function draw_playlist()
  -- ...
  playlist_osd_overlay.data = ass.text
  playlist_osd_overlay:update()
end

function remove_keybinds()
  playlist_osd_overlay.data = ""
  playlist_osd_overlay:update()
  -- ...
end

```

**Comment by verygoodlee** [comment]
see https://mpv.io/manual/stable/#command-interface-ass-events
>  If res_y is set to 0, PlayResY is initialized to an arbitrary default value (but note that the default for this command is 720, not 0). If res_x is set to 0, PlayResX is set based on res_y such that a virtual ASS pixel has a square pixel aspect ratio.

The key point is the `arbitrary default value`, and I finally found it in the source code.
https://github.com/mpv-player/mpv/blob/e22a2f04833852ce825eb7c1235d9bdaaa9b2397/sub/osd_libass.c#L111-L112
It's named `MP_ASS_FONT_PLAYRESY`, is a constant value `288`.
https://github.com/mpv-player/mpv/blob/e22a2f04833852ce825eb7c1235d9bdaaa9b2397/sub/ass_mp.h#L33

- `scale_playlist_by_window=yes` `showamount=auto` `style_ass_tags={\fs16}` `text_padding_y=10`
image

**Comment by verygoodlee** [comment]
For the issue of long filename, I think it's better to overflow to the right than to the bottom, we can use ASS Tag `\q2` to make long filenames overflow to the right. 
image

**Comment by jonniek** [comment]
> For the issue of long filename, I think it's better to overflow to the right than to the bottom, we can use ASS Tag \q2 to make long filenames overflow to the right.

Nice, that is probably a good default. I experimented a bit with a scrolling text, but it feels a bit jarring. I think overflowing the window makes more sense as a default.

I'll test the PR out a bit tomorrow and merge if I don't find any issues.

File: playlistmanager.lua [hunk 2]
```diff
@@ -237,6 +238,66 @@ if settings.system=="auto" then
...
+  msg.info('auto showamount: ' .. settings.showamount)
+  
+  -- no word wrapping, let long filenames overflow to the right
+  if settings.style_ass_tags == nil or settings.style_ass_tags == '' then
+    settings.style_ass_tags = '{\\q2}'
+  else
+    -- if wrapstyle tag exists, remove it
+    settings.style_ass_tags = settings.style_ass_tags:gsub('\\q%d+','')
+    -- add to end
+    settings.style_ass_tags = settings.style_ass_tags:gsub('}', '\\q2}')
```
Comment (jonniek) [inline playlistmanager.lua:295]: Why is this necessary? Would there be some issue if the tag is there twice?

I'm thinking maybe it's better not to modify the `style_ass_tags` at all, and just add the `\q2` to default styles.

Comment (verygoodlee) [reply playlistmanager.lua:295]: `\\q%d+` will match all original `\q` tags, remove original `\q` tag and add `\q2` tag to end, the `\q2` tag is very necessary. 
>{\fs20\q0} --remove--> {\fs20} --add--> {\fs20\q2}
{\fs20} --remove--> {\fs20} --add--> {\fs20\q2}

Comment (jonniek) [reply playlistmanager.lua:295]: I see there are different settings. I think I'll still change it to be \q2 in default styles, so it's user controlled rather than setting modifying it.

**Comment by jonniek** [comment]
Thanks for the PR! This is a great feature. I changed the default to be auto sized, and other configs a bit. Let me know if you think there are any issues in the changes I pushed.

**Comment by verygoodlee** [comment]
You replaced `mp.set_osd_ass(...)` by `mp.create_osd_overlay("ass-events")`, but no value is assigned to it's `res_y`, the default value is 720,  current calculation of `showamount` is incorrect, and it has caused a new isue.
- `curtain_opacity=0.7` screen dim size is baed on 360p
image

You need to change these two lines to `local h = 720`, 
https://github.com/jonniek/mpv-playlistmanager/blob/759b4076005f6df18d2717e8d695045709d55c05/playlistmanager.lua#L242
https://github.com/jonniek/mpv-playlistmanager/blob/759b4076005f6df18d2717e8d695045709d55c05/playlistmanager.lua#L651
and this comment need to modify, both them are based on 720p
https://github.com/jonniek/mpv-playlistmanager/blob/759b4076005f6df18d2717e8d695045709d55c05/playlistmanager.lua#L253

update: 
find an old issue,  `curtain_opacity=1` has no effect.
`settings.curtain_opacity <= 1.0`
https://github.com/jonniek/mpv-playlistmanager/blob/759b4076005f6df18d2717e8d695045709d55c05/playlistmanager.lua#L654

**Comment by jonniek** [comment]
Oh I see, I didn't notice any difference in rendering by setting the size, but I didn't test with the curtain.

I think it's fixed now, thanks!

**Comment by verygoodlee** [comment]
you pushed your test  conf `curtain_opacity=1.0` 🤣
